p2p/service/gossip_listener.go

```diff
@@ -67,6 +70,46 @@ func (l *Listener) Stop() {
 
 func (l *Listener) listenToGossip(ctx context.Context, dataHandler GossipDataHandler, gossipChannel chan GossipMessage, stop chan struct{}, channel string) {
 	l.WithContext(ctx).With().Info("start listening to gossip", log.String("protocol", channel))
+
+	// fill the channel with tokens to limit number of concurrent routines
+	tokenChan := make(chan struct{}, l.config.MaxGossipRoutines)
+	for i := 0; i < l.config.MaxGossipRoutines; i++ {
+		tokenChan <- struct{}{}
+	}
+
+	waitForGossipToken := func(ctx context.Context) {
+		// get a token to create a new channel
+		l.WithContext(ctx).With().Debug("waiting for available slot for gossip handler",
+			log.Int("available_slots", len(tokenChan)),
+			log.Int("total_slots", cap(tokenChan)))
+
+		if len(tokenChan) == 0 {
+			l.WithContext(ctx).Error("no available slots for gossip handler, blocking")
```

Comment:
do we need to notify user about this error? if not this is a warning

Reply:
maybe eventually this could be reduced to a warning, but for now, if this is happening in the devnet/testnet, we want to know about it right away since it will have ripple effects and may cause problems for downstream processes such as tortoise/hare